src/sage/graphs/homomorphisms/count_homomorphisms.py

```diff
@@ -0,0 +1,236 @@
+        #
+        # (node_index, bag_vertices) node_type
+        #
+        # Example: (5, {0, 4}) intro
+        self.dir_labelled_TD = label_nice_tree_decomposition(self.nice_tree_decomp, self.root, directed=True)
+
+        # `node_changes_dict` is responsible for recording introduced and
+        # forgotten vertices in a nice tree decomposition
+        self.node_changes_dict = node_changes(self.dir_labelled_TD)
+
+        # `DP_table` is a vector/list of dictionaries
+        # Each element (dict) corresponds to the (induced) hom's of a tree node.
+        # For each pair inside a dict, the key is the hom, and the value is the number of hom's:
+        #
+        # An example of K2 to K3, the values are arbitrary for demo purpose:
+        #
+        # [{((4, 0), (5, 1)): 10,
+        #   ((4, 0), (5, 2)): 20,
+        #   ((4, 1), (5, 0)): 30,
+        #   ((4, 1), (5, 2)): 40,
+        #   ((4, 2), (5, 0)): 50,
+        #   ((4, 2), (5, 1)): 60}, {}, ...]
+        self.DP_table = [{} for _ in range(len(self.dir_labelled_TD))]
+
+    def count_homomorphisms(self):
+        r"""
+        Return the number of homomorphisms from the graph `G` to the graph `H`.
+
+        A homomorphism from a graph `G` to a graph `H` is a function
+        `\varphi : V(G) \mapsto V(H)`, such that for any edge `uv \in E(G)` the
+        pair `\varphi(u) \varphi(v)` is an edge of `H`.
+
+        For more information, see the :wikipedia:`Graph_homomorphism`.
+
+        ALGORITHM:
+
+        This is an implementation based on the proof of Prop. 1.6 in [CDM2017]_.
+
+        OUTPUT:
+
+        - an integer, the number of homomorphisms from `graph` to `target_graph`
+
+        EXAMPLES::
+
+            sage: from sage.graphs.homomorphisms.count_homomorphisms import GraphHomomorphismCounter
+            sage: square = graphs.CycleGraph(4)
+            sage: bip = graphs.CompleteBipartiteGraph(2, 4)
+            sage: counter = GraphHomomorphismCounter(square, bip)
+            sage: counter.count_homomorphisms()
+            128
+        """
+        # Whether it's BFS or DFS, every node below join node(s) would be
+        # computed already, so we can go bottom-up safely.
+        for node in reversed(self.dir_labelled_TD.vertices()):
+            node_type = self.dir_labelled_TD.get_vertex(node)
+
+            match node_type:
+                case 'intro':
+                    self._add_intro_node(node)
+                case 'forget':
+                    self._add_forget_node(node)
+                case 'join':
+                    self._add_join_node(node)
+
+                case _:
+                    self._add_leaf_node(node)
+
+        return self.DP_table[0][0]
+
+    ### Main adding functions
+
+    def _add_leaf_node(self, node):
+        r"""
+        Add the computation result of processing leaf node(s) to the DP table.
+        """
+        node_index = get_node_index(node)
+        self.DP_table[node_index] = [1]
```

Comment:
lies 69-80, you explained that `DP_table` is a list of dictionary and here you assign a list instead of a dictionary. Please clarify the data structure you use and why such assignment might be valid.

Reply:
Sorry, old comment. Just corrected.